src/main/java/com/emc/object/s3/LargeFileUploader.java

```diff
@@ -138,84 +220,204 @@ public void upload() {
+            partsToSkip = resumeContext.getPartsToSkip();
+        }
 
-        List<Future<MultipartPartETag>> futures = new ArrayList<Future<MultipartPartETag>>();
+        if (uploadId == null) {
+            // initiate MP upload
+            uploadId = initMpu();
+        }
+
+        List<Future<MultipartPartETag>> futures = new ArrayList<>();
+        SortedSet<MultipartPartETag> parts = new TreeSet<>();
         try {
             // submit all upload tasks
-            int partNumber = 1;
-            long offset = 0, length = partSize;
-            while (offset < fullSize) {
+            int lastPart = (int) ((fullSize - 1) / partSize) + 1;
+            for (int partNumber = 1; partNumber <= lastPart; partNumber++) {
+                long offset = (partNumber - 1) * partSize;
+                long length = partSize;
                 if (offset + length > fullSize) length = fullSize - offset;
 
-                futures.add(executorService.submit(new UploadPartTask(uploadId, partNumber++, offset, length)));
-
-                offset += length;
+                // skip upload and reuse existing parts if found.
+                if (partsToSkip != null && partsToSkip.containsKey(partNumber)) {
+                    log.info("bucket {} key {} partNumber {} already exists, will be reused for multipart upload", bucket, key, partNumber);
+                    // re-read source part if necessary
+                    if (resumeContext.isVerifySkippedParts()) {
+                        futures.add(executorService.submit(new ReadSourcePartTask(partNumber, offset, length)));
+                    } else {
+                        parts.add(new MultipartPartETag(partNumber, partsToSkip.get(partNumber).getETag()));
+                    }
+                } else {
+                    futures.add(executorService.submit(new UploadPartTask(uploadId, partNumber, offset, length)));
+                }
             }
 
+            // TODO: allow calling code to stop the upload, allowing in-transfer parts to complete, but then shut down
+            //       and return completed part ETags (ResumeContext)
             // wait for threads to finish and gather parts
-            SortedSet<MultipartPartETag> parts = new TreeSet<MultipartPartETag>();
             for (Future<MultipartPartETag> future : futures) {
                 parts.add(future.get());
             }
 
             // complete MP upload
-            CompleteMultipartUploadRequest compRequest = new CompleteMultipartUploadRequest(bucket, key, uploadId).withParts(parts);
-            CompleteMultipartUploadResult result = s3Client.completeMultipartUpload(compRequest);
-            eTag = result.getETag();
+            eTag = completeMpu(uploadId, parts);
 
         } catch (Exception e) {
-
             // abort MP upload
             try {
-                s3Client.abortMultipartUpload(new AbortMultipartUploadRequest(bucket, key, uploadId));
+                abortMpu(uploadId);
```

Comment:
We probably shouldn't abort MPU if resume context is provided.

Reply:
this is handled by the new `abortMpuOnFailure` option